programs/voter-stake-registry/src/state/lockup.rs

```diff
@@ -75,42 +85,50 @@ impl Lockup {
     }
 
     /// Returns the number of periods left on the lockup.
+    /// Returns 0 after lockup has expired and periods_total before start_ts.
     pub fn periods_left(&self, curr_ts: i64) -> Result<u64> {
-        Ok(self
-            .periods_total()?
-            .saturating_sub(self.period_current(curr_ts)?))
-    }
-
-    /// Returns the current period in the vesting schedule.
-    pub fn period_current(&self, curr_ts: i64) -> Result<u64> {
         let period_secs = self.kind.period_secs();
         if period_secs == 0 {
             return Ok(0);
         }
         if curr_ts < self.start_ts {
-            return Ok(0);
+            return self.periods_total();
         }
-        let period = u64::try_from({
-            let secs_elapsed = curr_ts.saturating_sub(self.start_ts);
-            secs_elapsed.checked_div(period_secs).unwrap()
-        })
-        .map_err(|_| ErrorCode::UnableToConvert)?;
-        Ok(period)
+        Ok((self.seconds_left(curr_ts) + period_secs - 1) / period_secs)
+    }
+
+    /// Returns the current period in the vesting schedule.
+    /// Will report periods_total() after lockup has expired and 0 before start_ts.
+    pub fn period_current(&self, curr_ts: i64) -> Result<u64> {
+        Ok(self
+            .periods_total()?
+            .saturating_sub(self.periods_left(curr_ts)?))
     }
 
-    /// Returns the total amount of periods in the lockup period.
+    /// Returns the total amount of periods in the lockup.
     pub fn periods_total(&self) -> Result<u64> {
         let period_secs = self.kind.period_secs();
         if period_secs == 0 {
             return Ok(0);
         }
 
-        // Number of seconds in the entire lockup.
-        let lockup_secs = self.end_ts.checked_sub(self.start_ts).unwrap();
+        let lockup_secs = self.seconds_left(self.start_ts);
         require!(lockup_secs % period_secs == 0, InvalidLockupPeriod);
 
-        // Total periods in the entire lockup.
-        Ok(u64::try_from(lockup_secs.checked_div(period_secs).unwrap()).unwrap())
+        Ok(lockup_secs / period_secs)
+    }
+
+    /// Remove the vesting periods that are now in the past.
+    pub fn remove_past_periods(&mut self, curr_ts: i64) -> Result<()> {
+        let periods = self.period_current(curr_ts)?;
+        let period_secs = self.kind.period_secs();
+        self.start_ts = self
+            .start_ts
+            .checked_add(i64::try_from(periods.checked_mul(period_secs).unwrap()).unwrap())
+            .unwrap();
+        assert!(self.start_ts <= self.end_ts);
```

Comment:
nitpick: wouldnt require with error code be better than these 2 plain asserts, does assert return a Result?

Reply:
Asserts panic on failure, however, the release build doesn't even include them. They were asserts because they're documentation about assumptions. If the other code does what it should it should never fail.

However, given that the bug was exactly this assert failing, I'll make them regular checks instead. It would not be a user-correctable error though, more like an internal program error.